CPAC/registration/registration.py

```diff
@@ -3816,8 +3824,39 @@ def warp_timeseries_to_T1template_abcd(wf, cfg, strat_pool, pipe_num, opt=None
     node, out = strat_pool.get_data('from-T1w_to-template_mode-image_xfm')
     wf.connect(node, out, convert_func_to_standard_warp, 'warp2')
 
-    node, out = strat_pool.get_data('space-template_res-bold_desc-brain_T1w')
-    wf.connect(node, out, convert_func_to_standard_warp, 'reference')
+    if strat_pool.check_rpool('space-template_res-bold_desc-brain_T1w'):
+        node, out = strat_pool.get_data('space-template_res-bold_desc-brain_T1w')
+        wf.connect(node, out, convert_func_to_standard_warp, 'reference')
+    else:
+        anat_brain_to_func_res = pe.Node(interface=fsl.ApplyWarp(),
+                                    name=f'resample_anat_brain_in_standard_{pipe_num}')
+        anat_brain_to_func_res.inputs.interp = 'spline'
+        anat_brain_to_func_res.inputs.premat = cfg.registration_workflows[
+            'anatomical_registration']['registration']['FSL-FNIRT']['identity_matrix']
+
+        node, out = strat_pool.get_data('space-template_desc-preproc_T1w')
+        wf.connect(node, out, anat_brain_to_func_res, 'in_file')
+
+        node, out = strat_pool.get_data('T1w-template-funcreg')
+        wf.connect(node, out, anat_brain_to_func_res, 'ref_file')
+
+        wf.connect(anat_brain_to_func_res, 'out_file',
+                convert_func_to_standard_warp, 'reference')
+
+    # Create brain masks in this space from the FreeSurfer output (changing resolution)
+    # applywarp --rel --interp=nn -i ${FreeSurferBrainMask}.nii.gz -r ${WD}/${T1wImageFile}.${FinalfMRIResolution} --premat=$FSLDIR/etc/flirtsch/ident.mat -o ${WD}/${FreeSurferBrainMaskFile}.${FinalfMRIResolution}.nii.gz
+    anat_brain_mask_to_func_res = pe.Node(interface=fsl.ApplyWarp(),
+                                          name=f'resample_anat_brain_mask_in_standard_{pipe_num}')
+
+    anat_brain_mask_to_func_res.inputs.interp = 'nn'
+    anat_brain_mask_to_func_res.inputs.premat = cfg.registration_workflows[
+        'anatomical_registration']['registration']['FSL-FNIRT']['identity_matrix']
+
+    node, out = strat_pool.get_data('space-template_desc-T1w_mask')
+    wf.connect(node, out, anat_brain_mask_to_func_res, 'in_file')
+
+    wf.connect(anat_brain_to_func_res, 'out_file',
+               anat_brain_mask_to_func_res, 'ref_file')
```

Comment:
This is replicating https://github.com/FCP-INDI/C-PAC/blob/03a503c0370fa9915b9b841b0bd96e7eda3aa95b/CPAC/func_preproc/func_preproc.py#L1312-L1338?

As far as I can tell https://github.com/FCP-INDI/C-PAC/blob/03a503c0370fa9915b9b841b0bd96e7eda3aa95b/CPAC/func_preproc/func_preproc.py#L1296-L1306 is the only nodeblock that outputs `'space-template_res-bold_desc-brain_T1w'` and https://github.com/FCP-INDI/C-PAC/blob/03a503c0370fa9915b9b841b0bd96e7eda3aa95b/CPAC/registration/registration.py#L3716-L3748 is the only one that uses that resource as an input. Instead of duplicating the method, why don't we just make sure `bold_mask_anatomical_resampled` always runs (and runs before `warp_timeseries_to_T1template_abcd`) if `warp_timeseries_to_T1template_abcd` is going to run?

Reply:
Ya good idea @shnizzedy , Initially I planned to just get the missing resource `space-template_res-bold_desc-brain_T1w`, but then it got longer needing `space-template_desc-bold_mask` as well down the workflow.

Is there a better way than to check for one of these resource and calling the `nodeblock`? I mean like adding in any higher level functions?

Reply:
> Initially I planned to just get the missing resource `space-template_res-bold_desc-brain_T1w`, but then it got longer needing `space-template_desc-bold_mask` as well down the workflow.

Oh, cool, I'm glad the same existing nodeblock builds both!

> Is there a better way than to check for one of these resource and calling the nodeblock? I mean like adding "The call" in any higher level functions?

I'm not sure if it would work, but you might be able to do something like

```Python
    if strat_pool.check_rpool('space-template_res-bold_desc-brain_T1w'):
        outputs = {}
    else:
        wf, outputs = bold_mask_anatomical_resampled(wf, cfg, strat_pool, pipe_num, opt)
```

at the top of the `warp_timeseries_to_T1template_abcd` definition of and then

```Python
    outputs.update({
        'space-template_desc-preproc_bold': (extract_func_brain, 'out_file'),
        'space-template_desc-scout_bold': (extract_scout_brain, 'out_file'),
        'space-template_desc-head_bold': (merge_func_to_standard, 'merged_file')
    })
```

instead of assigning to outputs at the end.

Reply:
Awesome, what an elegant solution. I was wondering how to push the intermediate outputs to the `rpool` . 
Makes sense now. Thanks

Reply:
[Update]
Talked to @sgiavasis about abstracting `anat_brain_to_bold_res` and `anat_brain_mask_to_bold_res` as 2 separate sub-workflows from `bold_mask_anatomical_resampled` nodeblock, such that it can be reused (eg. here in `CPAC.registration.registration.warp_timeseries_to_T1template_abcd` nodeblock